src/libmeasurement_kit/ooni/meek_fronted_requests.cpp

```diff
@@ -0,0 +1,93 @@
+// information on the copying conditions.
+
+#include "../common/utils.hpp"
+#include "../ooni/constants.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+void meek_fronted_requests(std::string input, Settings options,
+                   Callback<Var<report::Entry>> callback,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    Var<Entry> entry(new Entry);
+
+    std::string expected_body, outer_host, inner_host;
+
+    if (options["expected_body"].empty()) {
+        expected_body = constants::MEEK_SERVER_RESPONSE;
+    } else {
+        expected_body = options["expected_body"];
+    }
+
+    if (!options["outer_host"].empty() &&
+        !options["inner_host"].empty()) {
+        outer_host = options["outer_host"];
+        inner_host = options["inner_host"];
+    } else {
+        std::list<std::string> outer_inner = split(input, ":");
+        if (outer_inner.size() != 2) {
+            logger->debug("Couldn't split input: %s", input.c_str());
+            callback(entry);
+            return;
+        }
+        // XXX: We should make sure that we remove leading and trailing whitespaces
+        outer_host = outer_inner.front();
+        inner_host = outer_inner.back();
+    }
+
+    // url parsing methods require a schema
+    outer_host.insert(0, "https://");
+    inner_host.insert(0, "https://");
+
+    ErrorOr<http::Url> outer_url = mk::http::parse_url_noexcept(outer_host);
+    ErrorOr<http::Url> inner_url = mk::http::parse_url_noexcept(inner_host);
+
+    if (!outer_url || !inner_url) {
+        logger->debug("Invalid url: '%s' or '%s'", outer_host.c_str(),
+                      inner_host.c_str());
+        callback(entry);
+        return;
+    }
+
+    (*entry)["input"] = {outer_url->address, inner_url->address};
+
+    options["http/url"] = "https://" + outer_url->address;
+    options["net/ssl"] = true;
+    http::Headers headers = { {"Host", inner_url->address} };
+    std::string body = ""; // spec says this is always a GET, so no body
+
+    logger->debug("Connecting to outer host %s and requesting inner url %s.",
+                  outer_url->address.c_str(),
+                  inner_url->address.c_str());
+
+    templates::http_request(entry, options, headers, body,
+                            [=](Error err, Var<http::Response> response) {
+                                if (err) {
+                                    logger->debug(
+                                        "meek_fronted_requests: http-request error: %s",
+                                        err.explain().c_str());
+
+                                    (*entry)["meek_fronted_requests"] =
+                                        err.as_ooni_error();
+                                }
+
+                                if (!response) {
+                                    logger->warn("null response");
+                                    callback(entry);
```

Comment:
Also: I'm not sure it's written anywhere, but: if there is no error, you can assume you have a valid response. In some cases, on error, the response is currently `nullptr`. I was thinking, however, to move to a situation where we guarantee that we never return `nullptr` (even if wrapped by `Var<>`), such that, if that happens, is a library bug causing an exception triggered by `Var<>` when you attempt to access the pointer, and one does not need to code to also consider the case in which it's null. Does this make sense?